### ISSUE TITLE
Multi-process data loading with iterable dataset

### ISSUE THREAD

**Issue by marcmk6** [pytorch/pytorch#77522]
### 🐛 Describe the bug

Hi, 

I'm implementing the multi-process data loading logic for my own `Iterable` dataset. However I observed a strange behavior while playing with the **second** example in the doc [here](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset), with implementing `worker_init_fn`.

The following is a code snippet to reproduce. `MyIterableDataset` and `worker_init_fn` are copied from the doc without any modification.

```python
import torch
import math
from torch.utils.data import DataLoader


class MyIterableDataset(torch.utils.data.IterableDataset):
    def __init__(self, start, end):
        super(MyIterableDataset).__init__()
        assert end > start, "this example code only works with end >= start"
        self.start = start
        self.end = end

    def __iter__(self):
        return iter(range(self.start, self.end))


# Define a `worker_init_fn` that configures each dataset copy differently
def worker_init_fn(worker_id):
    worker_info = torch.utils.data.get_worker_info()
    dataset = worker_info.dataset  # the dataset copy in this worker process
    overall_start = dataset.start
    overall_end = dataset.end
    # configure the dataset to only process the split workload
    per_worker = int(math.ceil((overall_end - overall_start) / float(worker_info.num_workers)))
    worker_id = worker_info.id
    dataset.start = overall_start + worker_id * per_worker
    dataset.end = min(dataset.start + per_worker, overall_end)


if __name__ == '__main__':
    ds = MyIterableDataset(start=0, end=500)

    dl = DataLoader(
        dataset=ds, batch_size=100, num_workers=2, worker_init_fn=worker_init_fn,
    )

    for e in dl:
        print(e.shape)
```

Running this snippet gives the result:
```
torch.Size([100])
torch.Size([100])
torch.Size([100])
torch.Size([100])
torch.Size([50])
torch.Size([50])
```
We can see that the last `100` examples are split into two batches of `50` examples. 

Similarly, when I changed `num_workers=3`, I got
```
torch.Size([100])
torch.Size([100])
torch.Size([100])
torch.Size([67])
torch.Size([67])
torch.Size([66])
```
I don't expect `num_workers` to affect the final batching results: there should be `5` batches of size `100` regardless of `num_workers`, so I guess this is a bug.

### Versions

```
Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.2 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.27.3)
CMake version: version 3.22.2
Libc version: N/A

Python version: 3.8.11 (default, Aug 16 2021, 12:04:33)  [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.2-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] pytorch-lightning==1.6.2
[pip3] torch==1.11.0
[pip3] torchmetrics==0.8.1
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] pytorch-lightning         1.6.2                    pypi_0    pypi
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torchmetrics              0.8.1                    pypi_0    pypi
```

**Comment by NicolasHug** [comment]
@marcmk6 thanks for opening this issue.

The discrepancy you observe can lead to terrible results when batchnorm is involved in models. I noticed the same issue in https://github.com/pytorch/data/issues/302 (it's related to datapipes, but still relevant for any kind of iterable dataset).

Is there any reason for closing your issue? I agree that this is worth raising awareness for this, and investigate potential fixes.

**Comment by marcmk6** [comment]
Hi Nicolas,

I also made a [post](https://discuss.pytorch.org/t/multi-process-data-loading-with-iterable-dataset/151771) in the PyTorch forum, where `ptrblck` gave me an explanation to this. 

After playing around with this, I can tell my [guess](https://discuss.pytorch.org/t/multi-process-data-loading-with-iterable-dataset/151771/3?u=mk6) is correct. 
And to avoid such incomplete batches, it's possible to adjust the `examples_per_worker` like this:
```python
batches_per_worker = (overall_end - overall_start) // batch_size // num_workers
examples_per_worker = batches_per_worker * batch_size
```
In this way, the first `num_workers - 1` workers load examples divisible by `batch_size` and these workers will not create incomplete batches. (The last worker may, however, depending on the total number of examples.)

So I guess this maybe an issue of documentation unclarity, rather than a bug.

**Comment by marcmk6** [comment]
Here's my implementation of the above solution.
```python
def worker_init_fn(_):
    worker_info = get_worker_info()
    dataset = worker_info.dataset  # the dataset copy in this worker process
    worker_id = worker_info.id
    batch_size = dataset.batch_size
    num_workers = worker_info.num_workers

    overall_start = 0
    overall_end = len(dataset)

    batches_per_worker = (overall_end - overall_start) // batch_size // num_workers
    examples_per_worker = batches_per_worker * batch_size

    worker_start = overall_start + worker_id * examples_per_worker
    if worker_id != num_workers - 1:
        # the first `num_workers - 1` workers load examples divisible by `batch_size`
        worker_end = min(worker_start + examples_per_worker, overall_end)
    else:
        # the last worker loads all remaining examples
        worker_end = overall_end

    # configure the dataset to only process the split workload

    """
    use `worker_start` and `worker_end` to configure iterables
    """
```
